### PR TITLE
Context Based Restrictions

### PR DESCRIPTION
**Pull request by zhenwan**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccXXX'

...
```
image
image
image
image
image
image
image
image

### CODE REVIEW

**Comment by zhenwan** [comment]
@kavya498  please ignore my old PR https://github.com/IBM-Cloud/terraform-provider-ibm/pull/3193. and only review this one.

File: ibm/config.go [hunk 1]
```diff
@@ -1299,6 +1311,32 @@ func (c *Config) ClientSession() (interface{}, error) {
...
+		cbrURL = fileFallBack(fileMap, c.Visibility, "IBMCLOUD_CONTEXT_BASED_RESTRICTIONS_ENDPOINT", c.Region, cbrURL)
+	}
+	contextBasedRestrictionsClientOptions := &contextbasedrestrictionsv1.Options{
+		Authenticator: authenticator,
+		URL:           envFallBack([]string{"IBMCLOUD_CONTEXT_BASED_RESTRICTIONS_ENDPOINT"}, cbrURL),
+	}
+
+	// Construct the service client.
+	session.contextBasedRestrictionsClient, err = contextbasedrestrictionsv1.NewContextBasedRestrictionsV1(contextBasedRestrictionsClientOptions)
+	if err == nil {
```
Comment (kavya498) [inline ibm/config.go:1329]: please check if client not nil condition as well
```
if err == nil && session.contextBasedRestrictionsClient != nil{
 	session.contextBasedRestrictionsClient.Service.EnableRetries(c.RetryCount, c.RetryDelay)
 	session.contextBasedRestrictionsClient.SetDefaultHeaders(gohttp.Header{
			"X-Original-User-Agent": {fmt.Sprintf("terraform-provider-ibm/%s", version.Version)},
		})
}
```

File: ibm/data_source_ibm_cbr_rule.go [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+	getRuleOptions.SetRuleID(d.Get("rule_id").(string))
+
+	rule, response, err := contextBasedRestrictionsClient.GetRuleWithContext(context, getRuleOptions)
+	if err != nil {
+		log.Printf("[DEBUG] GetRuleWithContext failed %s\n%s", err, response)
+		return diag.FromErr(fmt.Errorf("GetRuleWithContext failed %s\n%s", err, response))
+	}
+
+	d.SetId(fmt.Sprintf("%s", *getRuleOptions.RuleID))
+	if err = d.Set("id", rule.ID); err != nil {
```
[the remark this answers is not in the document]
Comment (zhenwan) [reply ibm/data_source_ibm_cbr_rule.go]: deleted

File: ibm/config.go (hunk 1 above)
Comment (zhenwan) [reply ibm/config.go]: checked

File: ibm/resource_ibm_cbr_rule.go [hunk 3]
```diff
@@ -0,0 +1,527 @@
...
+								},
+							},
+						},
+					},
+				},
+			},
+			"x_correlation_id": &schema.Schema{
+				Type:         schema.TypeString,
+				Computed:     true,
+				ValidateFunc: InvokeValidator("ibm_cbr_rule", "x_correlation_id"),
```
Comment (kavya498) [inline ibm/resource_ibm_cbr_rule.go:121]: computed attributes are API responses..
Validations is not required for computed attributes

File: ibm/resource_ibm_cbr_rule.go [hunk 4]
```diff
@@ -0,0 +1,527 @@
...
+				Description:  "The supplied or generated value of this header is logged for a request and repeated in a response header for the corresponding response. The same value is used for downstream requests and retries of those requests. If a value of this headers is not supplied in a request, the service generates a random (version 4) UUID.",
+			},
+			"transaction_id": &schema.Schema{
+				Type:         schema.TypeString,
+				Computed:     true,
+				ValidateFunc: InvokeValidator("ibm_cbr_rule", "transaction_id"),
```
Comment (kavya498) [inline ibm/resource_ibm_cbr_rule.go:127]: computed attributes are API responses..
Validations is not required for computed attributes

File: ibm/resource_ibm_cbr_zone.go [hunk 5]
```diff
@@ -0,0 +1,652 @@
...
+			"name": &schema.Schema{
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: InvokeValidator("ibm_cbr_zone", "name"),
+				Description:  "The name of the zone.",
+			},
+			"account_id": &schema.Schema{
+				Type:         schema.TypeString,
+				Computed:     true,
+				ValidateFunc: InvokeValidator("ibm_cbr_zone", "account_id"),
```
Comment (kavya498) [inline ibm/resource_ibm_cbr_zone.go:36]: computed attributes are API responses..
Validations is not required for computed attributes

File: ibm/resource_ibm_cbr_zone.go [hunk 6]
```diff
@@ -0,0 +1,652 @@
...
+			"x_correlation_id": &schema.Schema{
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: InvokeValidator("ibm_cbr_zone", "x_correlation_id"),
+				Description:  "The supplied or generated value of this header is logged for a request and repeated in a response header for the corresponding response. The same value is used for downstream requests and retries of those requests. If a value of this headers is not supplied in a request, the service generates a random (version 4) UUID.",
+			},
+			"transaction_id": &schema.Schema{
+				Type:         schema.TypeString,
+				Computed:     true,
+				ValidateFunc: InvokeValidator("ibm_cbr_zone", "transaction_id"),
```
Comment (kavya498) [inline ibm/resource_ibm_cbr_zone.go:152]: computed attributes are API responses..
Validations is not required for computed attributes

File: website/docs/d/cbr_rule.html.markdown [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+data "ibm_cbr_rule" "cbr_rule" {
+	rule_id = "rule_id"
+}
+```
+
+## Argument Reference
+
+Review the argument reference that you can specify for your data source.
+
+* `rule_id` - (Required, Forces new resource, String) The ID of a rule.
```
Comment (kavya498) [inline website/docs/d/cbr_rule.html.markdown:25]: Please remove `Forces new resource`, datasource ll not have any force new resource

File: website/docs/d/cbr_zone.html.markdown [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+data "ibm_cbr_zone" "cbr_zone" {
+	zone_id = "zone_id"
+}
+```
+
+## Argument Reference
+
+Review the argument reference that you can specify for your data source.
+
+* `zone_id` - (Required, Forces new resource, String) The ID of a zone.
```
Comment (kavya498) [inline website/docs/d/cbr_zone.html.markdown:25]: Please remove `Forces new resource`, datasource ll not have any force new resource

File: website/docs/r/cbr_rule.html.markdown [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+# ibm_cbr_rule
+
+Provides a resource for cbr_rule. This allows cbr_rule to be created, updated and deleted.
+
+## Example Usage
+
+```hcl
+resource "ibm_cbr_rule" "cbr_rule" { 
+	description = "this is an example of rule"
```
Comment (kavya498) [inline website/docs/r/cbr_rule.html.markdown:17]: This example doesnt have Required arguments of resource.. Is this a valid example?

File: website/docs/r/cbr_rule.html.markdown [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+}
+```
+
+## Argument Reference
+
+Review the argument reference that you can specify for your resource.
+
+* `contexts` - (Optional, List) The contexts this rule applies to.
```
Comment (kavya498) [inline website/docs/r/cbr_rule.html.markdown:25]: Please update docs according to resource schema..
contexts is marked as Optional in docs but is required in resource schema..

Can you please cross check all other arguments in both the resources?

File: website/docs/r/cbr_zone.html.markdown [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+		value = "169.23.22.0-169.23.22.255"
+	}
+}
+```
+
+## Argument Reference
+
+Review the argument reference that you can specify for your resource.
+
+* `addresses` - (Optional, List) The list of addresses in the zone.
```
Comment (kavya498) [inline website/docs/r/cbr_zone.html.markdown:30]: Please update docs according to resource schema..
contexts is marked as Optional in docs but is required in resource schema..

Can you please cross check all other arguments in both the resources?

**Comment by kavya498** [comment]
@zhenwan ,
please resolve conflicts

File: ibm/resource_ibm_cbr_rule.go (hunk 4 above)
Comment (zhenwan) [reply ibm/resource_ibm_cbr_rule.go:127]: removed

File: ibm/resource_ibm_cbr_zone.go (hunk 5 above)
Comment (zhenwan) [reply ibm/resource_ibm_cbr_zone.go:36]: removed

File: ibm/resource_ibm_cbr_zone.go (hunk 6 above)
Comment (zhenwan) [reply ibm/resource_ibm_cbr_zone.go:152]: removed

File: ibm/resource_ibm_cbr_rule.go (hunk 3 above)
Comment (zhenwan) [reply ibm/resource_ibm_cbr_rule.go:121]: removed "x_correlation_id" from schema

File: website/docs/d/cbr_rule.html.markdown (hunk 7 above)
Comment (zhenwan) [reply website/docs/d/cbr_rule.html.markdown:25]: removed Forces new resource

File: website/docs/d/cbr_zone.html.markdown (hunk 8 above)
Comment (zhenwan) [reply website/docs/d/cbr_zone.html.markdown:25]: removed Forces new resource,

File: website/docs/r/cbr_rule.html.markdown (hunk 9 above)
Comment (zhenwan) [reply website/docs/r/cbr_rule.html.markdown:17]: updated with a valid example

File: website/docs/r/cbr_rule.html.markdown (hunk 10 above)
Comment (zhenwan) [reply website/docs/r/cbr_rule.html.markdown:25]: changed to Required, and checked other arguments

File: website/docs/r/cbr_zone.html.markdown (hunk 11 above)
Comment (zhenwan) [reply website/docs/r/cbr_zone.html.markdown:30]: updated and checked

**Comment by zhenwan** [comment]
@kavya498 fixed according your review, can you please review again

**Pull request merged**